library/core/src/num/int_sqrt.rs

```diff
@@ -0,0 +1,285 @@
+///
+/// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+/// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+/// Encyclopedia."
+macro_rules! unsigned_fn {
+    ($UnsignedT:ident, $HalfBitsT:ident, $stages:ident) => {
+        /// Returns the [integer square root][1] of any
+        #[doc = concat!("[`", stringify!($UnsignedT), "`](prim@", stringify!($UnsignedT), ")")]
+        /// input.
+        ///
+        /// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+        /// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+        /// Encyclopedia."
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        pub const fn $UnsignedT(mut n: $UnsignedT) -> $UnsignedT {
+            if n <= <$HalfBitsT>::MAX as $UnsignedT {
+                $HalfBitsT(n as $HalfBitsT) as $UnsignedT
+            } else {
+                const EVEN_MAKING_BITMASK: u32 = !1;
+                let normalization_shift = n.leading_zeros() & EVEN_MAKING_BITMASK;
+                n <<= normalization_shift;
+
+                let s = $stages!(n);
+
+                let denormalization_shift = normalization_shift >> 1;
+                s >> denormalization_shift
+            }
+        }
+    };
+}
+
+/// Generates the first stage of the computation after normalization.
+macro_rules! first_stage {
+    ($original_bits:literal, $n:ident) => {{
+        const N_SHIFT: u32 = $original_bits - 8;
+        let n = $n >> N_SHIFT;
+
+        U8_ISQRT_WITH_REMAINDER[n as usize]
+    }};
+}
+
+/// Generates a middle stage of the computation.
+macro_rules! middle_stage {
+    ($original_bits:literal, $ty:ty, $n:ident, $s:ident, $r:ident) => {{
+        // SAFETY: Inform the optimizer that `$s` is nonzero. This will allow
+        // it to avoid generating code to handle division-by-zero panics in the
+        // divisions below.
+        //
+        // If the original `$n` is zero, the top of the `unsigned_fn` macro
+        // recurses instead of continuing to this point, so the original `$n`
+        // wasn't a 0 if we've reached here.
+        //
+        // Then the `unsigned_fn` macro normalizes `$n` so that at least one of
+        // the two most-significant bits is a 1.
+        //
+        // Then these stages take as many of the most-significant bits of `$n`
+        // as will fit in this stage's type. For example, the stage that
+        // handles `u32` deals with the 32 most-significant bits of `$n`. This
+        // means that each stage has at least one 1 bit in `n`'s two
+        // most-significant bits, making `n` nonzero.
+        //
+        // Then, the stage previous to this produces `$s` as the correct
+        // integer square root for the previous type. Since it was taking the
+        // integer square root of a nonzero number, `$s` will be nonzero.
+        unsafe { crate::hint::assert_unchecked($s != 0) };
```

Comment:
Could this move to the end of the stages where the precondition is satisfied, rather than the beginning of the stages where the precondition is used? Easier to keep track of the bounds that way, otherwise the macros should probably have a `# Safety` doc.

And to confirm, LLVM fails to elide the checks without these assertions?

Reply:
> Could this move to the end of the stages where the precondition is satisfied, rather than the beginning of the stages where the precondition is used? Easier to keep track of the bounds that way, otherwise the macros should probably have a `# Safety` doc.

It makes the code somewhat uglier, so I've added a `Safety` section to the macros' rustdoc comments.

> And to confirm, LLVM fails to elide the checks without these assertions?

Yes. Commenting out lines 202 and 241 on [the godbolt](https://rust.godbolt.org/z/nq6d13x7Y) introduces `core::panicking::panic_const::panic_const_div_by_zero` into the assembly code.

Reply:
Ooh - actually these could assert `s >= 0` right? You should try adding it to the end of `last_stage` too, then you might be able to remove the `assert_unchecked` calls from the API-public `sqrt` functions 🎉  (except for the MAX checks, but maybe that could even move to `unsigned_fn` / `signed_fn`)

Reply:
> You should try adding it to the end of last_stage too, then you might be able to remove the assert_unchecked calls from the API-public sqrt functions 

I'd have to keep it somewhere other than the `stages` section because `s >= 0` at the end of `last_stage` can't help in any way, as the value is stored in an unsigned integer at that time and `s >= 0` is redundant then.

> Except for the MAX checks, but maybe that could even move to `unsigned_fn` / `signed_fn` - and then the public `sqrt` functions don't need any unsafe code.

The main problem with that is that I have to do `const MAX_RESULT: u32 = core::num::int_sqrt::u32(u32::MAX);` or something like that to get a constant to use in `assert_unchecked`. That's fine everywhere except inside `core::num::int_sqrt::u32` because of infinite recursion. To keep the `assert_unchecked`s for `MAX` out of the public methods, I'd have to add a second layer of unsigned internal functions.

Reply:
I think that I can get it to work by moving the unsigned functions' contents to function names like `internal_16` and have both `i16` and `u16` call that. Inside `i16` and `u16` will also be the optimization hints. Is that a good idea?

Reply:
I'm not really worried about moving the MAX asserts out of public method, that was just a thought. Putting them at the end of the macros rather than the beginning is the more important thing so the algorithm-internal safety invariants (which are harder to verify) are more self-contained.

Reply:
OK, I've moved the asserts from `middle_stage!` and `last_stage!` into `first_stage!` and `middle_stage!`. Similar to before, with [the new godbolt](https://rust.godbolt.org/z/zeTjj8xfY), commenting out line 161 or 213 produces `core::panicking::panic_const::panic_const_div_by_zero` in the assembly code.

Reply:
I think I need to see the code to understand this haha, just push when you get the chance and I can close some of these out (doesn't matter if it's incomplete)

Reply:
OK, pushed.

Reply:
Thanks, this looks great now